jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/MetaInfConfiguration.java

```diff
@@ -95,166 +93,189 @@ public MetaInfConfiguration()
+        //noinspection UrlHashCode
+        Set<URL> metaInfTlds = (HashSet<URL>)context.getAttribute(METAINF_TLDS); // TODO: make this not a Set to avoid URL.hashcode issues
+        if (metaInfTlds == null)
+            metaInfTlds = new HashSet<>();
+        context.setAttribute(METAINF_TLDS, metaInfTlds);
+
+        Set<Resource> metaInfResources = (HashSet<Resource>)context.getAttribute(METAINF_RESOURCES);
+        if (metaInfResources == null)
+            metaInfResources = new HashSet<>();
+        context.setAttribute(METAINF_RESOURCES, metaInfResources);
+
+        Map<Resource, Resource> metaInfFragments = (HashMap<Resource, Resource>)context.getAttribute(METAINF_FRAGMENTS);
+        if (metaInfFragments == null)
+            metaInfFragments = new HashMap<>();
+        context.setAttribute(METAINF_FRAGMENTS, metaInfFragments);
+
+        // No pattern to apply for WEB-INF/classes, so just add to metadata
+        context.getMetaData().setWebInfClassesResources(findClassesDirs(context));
+
+        // Collect container paths that have selection patterns
+        // add them to the context metadata
+        List<Resource> containerResources = getContainerPaths(context);
+        containerResources.stream()
+            .forEach(r -> context.getMetaData().addContainerResource(r));
+
+        // Collect webapp paths that have selection patterns
+        // add them to the context metadata
+        List<Resource> webappResources = getWebAppPaths(context);
+        webappResources.stream()
+            .forEach(r -> context.getMetaData().addWebInfResource(r));
+
+        // -- Scan of META-INF directories --
+
+        // Figure out container caching rules
+        boolean useContainerCache = getUseContainerCache(context);
+        Server server = context.getServer();
+        Map<Resource, Resource> metaInfResourceCache = getResourceCache(server, useContainerCache);
+        Map<Resource, Resource> metaInfFragmentCache = getFragmentCache(server, useContainerCache);
+        Map<Resource, Collection<URL>> metaInfTldCache = getTldCache(server, useContainerCache);
 
-        //find web-app jars and select which ones to scan
-        findAndFilterWebAppPaths(context);
+        // Restrict scan to resource targets that have a META-INF directory
+        List<Resource> containerTargets = containerResources.stream()
+            .filter(r -> Resources.isReadableDirectory(r.resolve("META-INF")))
+            .toList();
 
-        //No pattern to appy to classes, just add to metadata
-        context.getMetaData().setWebInfClassesResources(findClassDirs(context));
+        List<Resource> webappTargets = webappResources.stream()
+            .filter(r -> Resources.isReadableDirectory(r.resolve("META-INF")))
+            .toList();
 
-        scanJars(context);
+        // Scan for META-INF/*.tld entries
+        metaInfTlds.addAll(scanTlds(streamTargets(containerTargets, webappTargets, true), metaInfTldCache));
```

Comment:
I think the idea was not to iterate over the resources multiple times, but iterate just once and interrogate for tlds, fragments, resources in one go.